### PR TITLE
feat: goal-diffs

### LINKED ISSUE #1469: Highlight changes in goal state
tl;dr: https://twitter.com/famontesi/status/1557346899610443782

We already have both the state before and after a tactic in the info tree, we should use them at the same time! A full diff is probably too noisy since the old state is usually less interesting than the new one, as long as you can see where it changed at all. We could probably do this by diffing the `TaggedText`s, introducing a new kind of `MsgEmbed`. The Coq implementation is documented at https://coq.inria.fr/distrib/current/refman/proofs/writing-proofs/proof-mode.html#how-diffs-are-calculated.

### PR DESCRIPTION
**Pull request by EdAyers**
goal-diff-final

`InteractiveGoals` now has information that a client infoview can use to show what parts of the goal have changed after applying a tactic.

There are lots of exciting ways to extend this, but I think this PR should be just limited to a first version and getting a sensible and extensible server/client API in place.

## API changes

All API changes are conservative so old versions of vscode will still be able to render the goals.

- `InteractiveGoals`, `InteractiveGoal`, `InteractiveHypothesisBundle` now have a `message? : Option String` field that can be used to emit debugging messages.
- `InteractiveGoal` and `InteractiveHypothesisBundle` now have a `isInserted? : Option Bool` which is true when it is a new goal or hypothesis and should be rendered with a green highlight.
- `SubexprInfo` has a new `tags?: Option String` field. Currently the infoview is highlighting green anything that has the tag `inserted` on it. I would like some comments on whether this is a good way to implement the diff api. The tag approach could cause issues later because it depends on both client and server using the same magic string. Another problem is that it might be causing bloat of `SubexprInfo`, perhaps instead the diff info should live on its own map from position to tags.

## Todos

- [x] tidy up
- [x] ~~#1607~~ no longer depends on this
- [ ] vscode extension PR https://github.com/leanprover/vscode-lean4/pull/251
- [x] unit tests
- [x] explain the api here
- [x] write a docstring explaining diffing algorithm used

## Todos/ future work

- [x] We use `goalsBefore` and `goalsAfter` on the `TacticInfo` object to get the diff. Currently I only show the diff on `useAfter = true`, perhaps I can also write a version for `useAfter = false` so that you see the change that is _about_ to happen.
- [ ] If a hypothesis is deleted, render it with a red triangle at the point that the hypothesis was deleted.

Resolves #1469

### CODE REVIEW

File: src/Lean/Util/FindExpr.lean
Comment (Kha) [inline src/Lean/Util/FindExpr.lean:11]: What does "immediate" mean here?

File: src/Lean/SubExpr.lean
Comment (Kha) [inline src/Lean/SubExpr.lean:237]: Are any of these functions actually used?

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean:121]: yes still working on that case

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean:40]: I don't really understand how the delayed mvars work in lean4. I was worried that there would be a way of getting an infinite descent with `getDelayedMVarRoot m` so I added a fuel param to the recursion.

The question is:

Let `m₁ ⊏ m₂` when `m₁ ≠ m₂` and `m₂ = getDelayedMVarRoot m₁` or `m₂` occurs in `m₁.assignment`. If I take the transitive closure of `⊏`, is it still irreflexive? It feels like it ought to be but I wasn't sure.

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean:52]: Yes I'm not sure on what names to use for these yet. I'm interpreting `inserted` as having the semantics of "paint it green".

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean:128]: Yeah I just thought it looked a little better but very subjective

File: src/Lean/Util/FindExpr.lean
Comment (EdAyers) [reply src/Lean/Util/FindExpr.lean]: added some clarification

File: src/Lean/SubExpr.lean
Comment (EdAyers) [reply src/Lean/SubExpr.lean]: removed, an earlier more complicated version of diff used them

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean:52]: I'll rename to `change`

File: src/Lean/Widget/Diff.lean
Comment (Kha) [reply src/Lean/Widget/Diff.lean:168]: yes

File: src/Lean/Widget/Diff.lean
Comment (Kha) [reply src/Lean/Widget/Diff.lean:40]: I don't know whether searching through mvar assignments is the right thing to do here. Note that tactics could produce an arbitrarily long chain between an input and output mvar, but we definitely don't want to do an arbitrarily deep search through the proof term.

Comment (EdAyers) [reply src/Lean/Widget/Diff.lean]: Ok I've removed the recursion and it seems to work ok on my set of examples

Comment (EdAyers) [reply src/Lean/Widget/Diff.lean]: Ah no it doesn't work on `rw [h₁] at h₂`, you need to recurse to get the parent relationship. I am a bit worried about using a heuristic where you match up goals in zip order because it seems to me it's going to be easy to produce misleading output, particularly if the tactic is being applied to many goals with a combinator.

Comment (EdAyers) [reply src/Lean/Widget/Diff.lean]: I'll add a cache so at least it doesn't check the same mvar over and over.

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean]: I've removed this for now.

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean]: This should now diff lambdas and pis correctly.

File: src/Lean/Widget/InteractiveGoal.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveGoal.lean:52]: If the `message?` fields are only used for development, do we need to keep them around? What's the advantage of sending back a `message` rather than throwing an exception (which would also show up in the infoview)?

Comment (EdAyers) [reply src/Lean/Widget/InteractiveGoal.lean:52]: They are just for development. It's harder to debug if I used exceptions because I wouldn't be able to look at the tactic state.

Comment (EdAyers) [reply src/Lean/Widget/InteractiveGoal.lean:52]: I will remove at the end

File: src/Init/Data/List/Basic.lean
Comment (Vtec234) [inline src/Init/Data/List/Basic.lean:460]: `isSuffix?` might be a little more consistent with the rest of the library (e.g. `Syntax.isNatLit?`).

File: src/Init/Data/Option/Basic.lean
Comment (Vtec234) [inline src/Init/Data/Option/Basic.lean:80]:
```suggestion
Otherwise act like `orElse`. -/
```

File: src/Lean/Widget/InteractiveGoal.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveGoal.lean:25]:
```suggestion
  /-- The value, in the case that the hypothesis is a `let`-binder. -/
```

File: src/Lean/Widget/InteractiveGoal.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveGoal.lean:32]: This has three possible values. What is the semantics of `some false`? If the `Option` is just to avoid sending extra data when unnecessary, let's explain in the comment that `some false` should never be used.

File: src/Lean/Widget/InteractiveGoal.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveGoal.lean:87]:
```suggestion
/-- This is everything needed to render an interactive term goal in the infoview. -/
```

File: src/Lean/Widget/InteractiveGoal.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveGoal.lean:131]: Should this be here or somewhere in `Meta`?

File: src/Lean/Widget/InteractiveCode.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveCode.lean:35]:
```suggestion
```
This function is short enough and a little difficult to explain (I don't understand the docstring at all). I would just remove the docstring.

File: src/Lean/Widget/Diff.lean
Comment (Vtec234) [inline src/Lean/Widget/Diff.lean:75]:
```suggestion
  describing how the expression has changed relative to `before` at the given position.-/
```
Do I understand correctly that this is what the map describes?

File: src/Init/Data/List/Basic.lean
Comment (EdAyers) [reply src/Init/Data/List/Basic.lean:460]: on the other hand, this would mean that `isPrefix` and `isPrefix?` would be doing different things which might be confusing. I was hoping there would be a fn in mathlib3 I could steal the name from.

File: src/Lean/Widget/InteractiveGoal.lean
Comment (EdAyers) [reply src/Lean/Widget/InteractiveGoal.lean:32]: If I do `isInserted? : Bool`, will the json deriver encode that properly?

File: src/Lean/Widget/InteractiveGoal.lean
Comment (EdAyers) [reply src/Lean/Widget/InteractiveGoal.lean:131]: I wasn't sure, it's an extraction of functionality of the method below, and it appeared to me to be doing an InteractiveGoal-specific thing of `sanitizeNames` which I didn't really understand so I kept it here.

File: src/Lean/Widget/InteractiveGoal.lean
Comment (Vtec234) [reply src/Lean/Widget/InteractiveGoal.lean:32]: No, we'd be sending `"isInserted": false` then which takes up space, so the `Option` seems like the right choice; but it seems worth noting that this is the reason for using `Option`.

**Comment by Vtec234** [comment]
> * `SubexprInfo` has a new `tags?: Option String` field. Currently the infoview is highlighting green anything that has the tag `inserted` on it. I would like some comments on whether this is a good way to implement the diff api. The tag approach could cause issues later because it depends on both client and server using the same magic string. Another problem is that it might be causing bloat of `SubexprInfo`, perhaps instead the diff info should live on its own map from position to tags.

From the perspective of the size of sent packets this is not an issue because we serialize `tags? := none` as nothing (0 bytes) in JSON. Earlier on I was thinking we may still prefer to use an external map `SubexprPos -> Tag` alongside the current `TaggedText<SubexprInfo>` rather than including `Tag`s in the `SubexprInfo`, basically because the current set of tags (used for goal diffs) is only relevant to the goal view; so when we use `CodeWithInfos` aka `TaggedText<SubexprInfo>` elswhere, such as in error messages, we get the "extra stuff" of tags which could be considered bloat.

*However* I am now thinking that tags could be useful more generally. For example, some error messages may want to bring the user's attention to a subexpression at which a tactic failed, or whatever, while still displaying the full top-level expression. For this though, it might be better to think of the infoview as a "dumb" display which doesn't know the semantics of "inserted/deleted/relevant" subexpression but simply gives it a particular colour or style. So i'd suggest renaming the tags to something style-y rather than semantic-y. E.g. "green"/"read". What do you think?

File: src/Lean/Widget/Diff.lean
Comment (EdAyers) [reply src/Lean/Widget/Diff.lean]: Fixed: I should have been using `Lean.Meta.getMVars`.

File: src/Init/Data/List/Basic.lean
Comment (EdAyers) [reply src/Init/Data/List/Basic.lean:460]: @digama0 what should it be called?

**Comment by EdAyers** [comment]
Yes I'm always in favour of functional styling

File: src/Init/Data/List/Basic.lean
Comment (digama0) [reply src/Init/Data/List/Basic.lean:460]: What do you mean by them doing different things? We do have a naming convention that `?` means Option-returning, and it is true that `(l1.isPrefix? l2).isSome == l1.isPrefix l2` so it seems okay to me.

...Oh, are prefix and suffix switched in this definition? I would recommend using `l1.isPrefix? l2` for the function that returns t such that `l2 ++ t == l1`, and similarly for `isSuffix?`.

File: src/Lean/Widget/InteractiveGoal.lean
Comment (EdAyers) [reply src/Lean/Widget/InteractiveGoal.lean]: removed

File: src/Lean/Widget/InteractiveCode.lean
Comment (Vtec234) [inline src/Lean/Widget/InteractiveCode.lean:42]: I think `deriving FromJson, ToJson` should have the same effect, although you'd still need the manual `HighlightColor.toString`.

**Review by Vtec234 — APPROVED** [review]
I didn't review `Diff.lean` in detail but everything else looks good!

File: src/Lean/Widget/InteractiveCode.lean
Comment (EdAyers) [reply src/Lean/Widget/InteractiveCode.lean:42]: ah I didn't realise deriving worked on enums now :)

Comment (EdAyers) [reply src/Lean/Widget/InteractiveCode.lean:42]: I've just deleted the toJson fromJson becuase it's not sent to the client anyway

File: src/Lean/Expr.lean
Comment (Kha) [inline src/Lean/Expr.lean:869]: This is a somewhat dangerous operation as it can expose loose bvars. I would suggest to merge it into `getForallBinderNames`, and perhaps to move all of these to the use site as they are quite usage specific.

**Review by Kha — COMMENTED** [review]
I didn't think about the new heuristics too hard yet. I'd say let's simply give them a try.

File: src/Lean/Widget/InteractiveCode.lean
Comment (Vtec234) [reply src/Lean/Widget/InteractiveCode.lean:42]: We still have the manual instances for `HighlightColor`

Comment (EdAyers) [reply src/Lean/Widget/InteractiveCode.lean:42]: oops

**Comment by Kha** [comment]
@EdAyers Feel free to PR a changelog entry for RELEASES.md once the extension change has landed!

**Comment by gebner** [comment]
Sorry for only getting to this PR now.

> "green background" may have different meanings, it might be better to think of the infoview as a "dumb" display which doesn't know the semantics of "inserted/deleted/relevant" subexpression but simply gives it a particular colour or style. So i'd suggest renaming the tags to something style-y rather than semantic-y. E.g. "green"/"red". What do you think?

For the record, I would have preferred "inserted" and "removed" instead of "green" and "red," resp.  There are several reasons:
 1. It is consistent with the `isInserted?` and `isRemoved?` fields.
 2. We could maybe show an "inserted" tooltip to explain what the color means.  (I didn't think  about this a lot though, the tooltip probably doesn't mix well with the popup we already have.)
 3. The "blue" color clashes with the highlighting.

**Comment by digama0** [comment]
Also, it's generally a bad idea to talk about colors directly because people have themes that can make the colors different, and then it's really confusing when `isGreen` means it's blue and `isRed` means it's gray in the theme. (Also colorblind people exist, and they usually switch the themes to something appropriate.)

**Comment by Vtec234** [comment]
I agree, but it seems none of the points raised here apply to the style-predicates vs semantic-predicates choice, it's just that `green/red/blue` is maybe not the best choice of style-predicates. To be fair, we could think of `inserted/removed` as style-predicates which indicate "draw this in the same way `git diff` would".

**Comment by EdAyers** [comment]
I am happy to open a PR reverting to semantic tags instead of colours. In response to the theming and colour-blindness, the actual colours that are displayed at the moment are a function of whether the theme is light or dark. It should be possible to get the colours to be completely theme-dependent by using [these](https://code.visualstudio.com/api/references/theme-color#integrated-terminal-colors).

**Comment by Kha** [comment]
> In the case of using a semantic tags model, are you ok with having a `tags?: string[]` field or would you prefer something more type-safe like `diffTag?: 'inserted' | 'removed'`?

Let me highlight this question since people may have missed the edit. Personally I don't have any strong opinions on it.
I'm assuming we're waiting on this change before the extension PR is merged.

**Pull request merged**
